lib/compiler/src/beam_ssa_type.erl

```diff
@@ -17,75 +17,383 @@
-      Linear :: [{non_neg_integer(), beam_ssa:b_blk()}],
-      Args :: [beam_ssa:b_var()],
-      Anno :: beam_ssa:anno(),
-      FuncDb :: func_info_db().
-opt_start(Linear, Args, Anno, FuncDb) ->
-    %% This is the first run through the module, so our arg_types can be
-    %% incomplete as we may not have visited all call sites at least once.
-    Ts = maps:from_list([{V,any} || #b_var{}=V <- Args]),
-    opt_continue_1(Linear, Args, get_func_id(Anno), Ts, FuncDb).
+%%
+
+-spec opt_start(term(), term()) -> term().
+opt_start(StMap, FuncDb0) when FuncDb0 =/= #{} ->
+    {ArgDb, FuncDb} = signatures(StMap, FuncDb0),
+
+    opt_start_1(maps:keys(StMap), ArgDb, StMap, FuncDb);
+opt_start(StMap, FuncDb) ->
+    %% Module-level analysis is disabled, likely because of a call to
+    %% load_nif/2 or similar. opt_continue/4 will assume that all arguments and
+    %% return types are 'any'.
+    {StMap, FuncDb}.
+
+opt_start_1([Id | Ids], ArgDb, StMap0, FuncDb0) ->
+    case ArgDb of
+        #{ Id := ArgTypes } ->
+            #opt_st{ssa=Linear0,args=Args} = St0 = map_get(Id, StMap0),
+
+            Ts = maps:from_list(zip(Args, ArgTypes)),
+            {Linear, FuncDb} = opt_function(Linear0, Args, Id, Ts, FuncDb0),
+
+            St = St0#opt_st{ssa=Linear},
+            StMap = StMap0#{ Id := St },
+
+            opt_start_1(Ids, ArgDb, StMap, FuncDb);
+        #{} ->
+            %% Unreachable functions must be removed so that opt_continue/4
+            %% won't process them and potentially taint the argument types of
+            %% other functions.
+            StMap = maps:remove(Id, StMap0),
+            FuncDb = maps:remove(Id, FuncDb0),
+
+            opt_start_1(Ids, ArgDb, StMap, FuncDb)
+    end;
+opt_start_1([], _CommittedArgs, StMap, FuncDb) ->
+    {StMap, FuncDb}.
+
+%%
+%% The initial signature analysis is based on the paper "Practical Type
+%% Inference Based on Success Typings" [1] by `Tobias Lindahl` and
+%% `Konstantinos Sagonas`, mainly section 6.1 and onwards.
+%%
+%% The general idea is to start out at the module's entry points and propagate
+%% types to the functions we call. The argument types of all exported functions
+%% start out a 'any', whereas local functions start at 'none'. Every time a
+%% function call widens the argument types, we analyze the callee again and
+%% propagate its return types to the callers, analyzing them again, and
+%% continuing this process until all arguments and return types have been
+%% widened as far as they can be.
+%%
+%% Note that we do not "jump-start the analysis" by first determining success
+%% types as in the paper because we need to know all possible inputs including
+%% those that will not return.
+%%
+%% [1] http://www.it.uu.se/research/group/hipe/papers/succ_types.pdf
+%%
+
+-record(worklist,
+        { counter = 0 :: integer(),
+          elements = gb_trees:empty() :: gb_trees:tree(integer(), term()),
+          indexes = #{} :: #{ term() => integer() } }).
```

Comment:
For clarity, I suggest making #worklist{} an abstract type. (It is *almost* an abstract type already.) That is, move the `worklist` record declaration to just before the implementation of the functions that operate on it, add a new `wl_new/0` function, and define a `worklist()` type. It is allowed to reference a type that is being defined later in the same module, so you can use the `worklist()` type for the `wl` field in `#sig_st{}`.

Reply:
Sure thing, fixed locally.